### PR TITLE
Refactor QA setup

### PR DESCRIPTION
**Pull request by aleksihakli**
Fixes jazzband-roadies/help#182

Some notable changes:
- Deprecate explicit support for Python 3.4 and 3.5 in order to simplify the test matrix.
- Clean up the test runner setup with tox and Travis

### CODE REVIEW

**Comment by jezdez** [comment]
Thanks @aleksihakli, but this needs a Python environment list still, check out the constance config: https://github.com/jazzband/django-constance/blob/master/.travis.yml

**Comment by aleksihakli** [comment]
@jezdez Python version list is now included in `.travis.yml`.

**Comment by jezdez** [comment]
I'm sorry, I wasn't clear about what I mean, tox-travis is best set up in using per-Python version test harnesses like in https://github.com/jazzband/django-constance/blob/b62206da573c1447b79ebc95a430df809672092d/.travis.yml#L3-L8.

**Comment by aleksihakli** [comment]
@jezdez I made an intermittent commit with the Python version listing as usual in Django projects but it won't bring much joy as the test matrix involved is more complex than a pair of `(Python, Django)` requirements.

It also includes another 3rd party packages such as Pillow or Boto and the matrix configuration then becomes `(Python, Django, Package)`.

Thus the the test matrix definition is essentially the same size, but the Python variables will be introduced twice in the `python` definition as well as `matrix` definition on the top level of `.travis.yml`.

The problem is that there needs to be a way to specify the Django AND package versions along with the Python version with `tox-travis` or otherwise the matrix configuration can not be simplified that much if we want to test the packages as defined here.

It would of course be possible to define multiple environment variables where one is e.g. `DJANGO` and another is `PACKAGE`, but manually including or excluding test matrix cells in such a configuration creates essentially the same structure.

**Comment by aleksihakli** [comment]
Just as a clarification: using the `python` keyword and the `jobs.include` (or `matrix.include`) will define both the Python versions and the additional matrix elements, so no joy there:

image

**Comment by aleksihakli** [comment]
And continuing on the testing subject, specifying both e.g. `DJANGO` and `PACKAGE` variables leads to a confusing matrix as well, so specifying Python version and TOXENV does seem like a good approach, although using tox-travis as well is not a bad idea.

image

File: .travis.yml [hunk 1]
```diff
@@ -1,138 +1,61 @@
...
     packages:
       - libgraphicsmagick1-dev
-      # - libgraphicsmagick++1-dev
-      # - libboost-python-dev
       - graphicsmagick
       - libjpeg62
       - zlib1g-dev
-
 cache:
-  apt: true
```
Comment (aleksihakli) [inline .travis.yml]: apt parameter is deprecated

File: .travis.yml [hunk 2]
```diff
@@ -1,138 +1,61 @@
...
+  - DJANGO=3.0  PACKAGE=imagemagick
+  - DJANGO=3.0  PACKAGE=graphicsmagick
+  - DJANGO=3.0  PACKAGE=redis
+  - DJANGO=3.0  PACKAGE=wand
+  - DJANGO=3.0  PACKAGE=dbm
 addons:
   apt:
     packages:
       - libgraphicsmagick1-dev
-      # - libgraphicsmagick++1-dev
```
Comment (aleksihakli) [inline .travis.yml]: these are available in git history

File: .travis.yml [hunk 3]
```diff
@@ -1,138 +1,61 @@
...
   email: false
-
 deploy:
   provider: pypi
   user: jazzband
   server: https://jazzband.co/projects/sorl-thumbnail/upload
   distributions: sdist bdist_wheel
   password:
     secure: BZlzJyqEbo1nofxEqeE12EyIcM6ldJL40yDRo62qk1dBirUV+eJQN3kTJtKTqVgTMRalQJKqMZgbhWtvxo2wsuBx+JW3LlfjYuY/fNzj/+5x7Tq/NEy4HyBj4RdWdincmdvQ0n4ixar6cDr79KjHUzC9Db3psbTRE6+Vw0EJTM4=
+  skip_existing: true
```
Comment (aleksihakli) [inline .travis.yml]: this prevents multiple conflicting deployments

File: .travis.yml [hunk 4]
```diff
@@ -1,138 +1,61 @@
...
-    - python: "3.8"
-      env: TOXENV=py38-django30-wand
-    - python: "3.8"
-      env: TOXENV=py38-django30-dbm
-
+python:
+  - 3.6
+  - 3.7
+  - 3.8
+env:
```
Comment (aleksihakli) [inline .travis.yml]: specifying the matrix this way runs tests on all supported environment combinations

File: setup.py [hunk 5]
```diff
@@ -1,4 +1,3 @@
-# -*- encoding: utf8 -*-
```
Comment (aleksihakli) [inline setup.py]: this is unnecessary with Python 2.7 deprecation

File: setup.py [hunk 6]
```diff
@@ -33,17 +32,13 @@ def run(self):
         'License :: OSI Approved :: BSD License',
         'Operating System :: OS Independent',
         'Programming Language :: Python',
-        'Programming Language :: Python :: 3.4',
```
Comment (aleksihakli) [inline setup.py]: Python 3.4 is EOL

File: setup.py [hunk 7]
```diff
@@ -33,17 +32,13 @@ def run(self):
...
-        'Programming Language :: Python :: 3.5',
```
Comment (aleksihakli) [inline setup.py]: supporting Python 3.5 is more effort than benefit; decreasing users, Django has dropped support in new versions

File: setup.py [hunk 8]
```diff
@@ -33,17 +32,13 @@ def run(self):
...
         'Programming Language :: Python :: 3.6',
         'Programming Language :: Python :: 3.7',
         'Programming Language :: Python :: 3.8',
         'Topic :: Internet :: WWW/HTTP :: Dynamic Content',
         'Topic :: Multimedia :: Graphics',
         'Framework :: Django',
         'Framework :: Django :: 1.11',
-        'Framework :: Django :: 2.0',
```
Comment (aleksihakli) [inline setup.py]: Django 2.0 is EOL

File: setup.py [hunk 9]
```diff
@@ -33,17 +32,13 @@ def run(self):
...
-        'Framework :: Django :: 2.1',
```
Comment (aleksihakli) [inline setup.py]: Django 2.1 is EOL as well

File: sorl/thumbnail/engines/pil_engine.py [hunk 10]
```diff
@@ -277,7 +277,7 @@ def _get_raw_data(self, image, format_, quality, image_info=None, progressive=Fa
             # Do not save unnecessary exif data for smaller thumbnail size
             params.pop('exif', {})
             image.save(bf, **params)
-        except (IOError, OSError):
```
Comment (aleksihakli) [inline sorl/thumbnail/engines/pil_engine.py]: Python 3 aliased IOError to OSError so they are essentially the same exception

flake8 threw a warning on this

File: tox.ini [hunk 11]
```diff
@@ -1,48 +1,62 @@
...
+commands =
+  py.test -rw --cov-config ../setup.cfg --cov sorl
 
 [testenv:flake8]
 changedir = {toxinidir}
 skip_install = True
 deps =
   flake8
   flake8-bugbear
-  # flake8-import-order
```
Comment (aleksihakli) [inline tox.ini]: available in git history as well

File: .travis.yml [hunk 12]
```diff
@@ -1,138 +1,68 @@
...
-    - python: "3.8"
-      env: TOXENV=py38-django30-graphicsmagick
-    - python: "3.8"
-      env: TOXENV=py38-django30-redis
-    - python: "3.8"
-      env: TOXENV=py38-django30-wand
-    - python: "3.8"
-      env: TOXENV=py38-django30-dbm
-
+    - stage: deploy
```
Comment (aleksihakli) [inline .travis.yml]: Refer to https://travis-ci.org/travis-ci/travis-build/builds/552914301/config

**Comment by aleksihakli** [comment]
@jezdez I think the configuration is more to your liking now, figured out how to configure the matrix in a way that should fit this project and the requirements. Also added `setuptools_scm` to the mix to reduce the burden of remembering to update the `sorl.__version__` for every release separately.

@camilonova the test bench should work now and only do one deployment per tag utilizing the Travis job stages. The setups should also auto-version now as long as the git tags are in place :)

I think that it's best to explicitly exclude combinations with `jobs.exclude` or `matrix.exclude` from this configuration if necessary, but otherwise test all the combinations of Python, Django and 3rd party tool versions, as comprehensive testing costs very little and makes me happy, but mistakes might incite the frowning of package users.

**Comment by aleksihakli** [comment]
@jezdez @camilonova the work in this PR is now ready for upstreaming work-wise.

There is a lot of changes in the one PR but they all relate to CI, testing, and versioning, and splitting them up doesn't simplify things, as it's easiest to see and review their correlation when implemented in one patch.

All the changes aside from the `setuptools_scm` and versioning tune-up are required or nice-to-have for a functioning Python 3 build and tests for the project and the docs.

**Comment by aleksihakli** [comment]
@jezdez @camilonova the work in this PR is now ready for review and possible upstreaming work-wise, I'm done with iterating over it.

There are a lot of changes in this one PR and patch, but they all relate to CI, testing, and versioning. Splitting them up doesn't really simplify things, as it's easiest to see and review their correlation at once. 

All the changes aside from the `setuptools_scm` and versioning tune-up are required or nice-to-have for a functioning Python 3 test and build setup for the project.

The docs were also broken on Python 3 due to the use of `xrange`, and I decided it would be nice to fix them in the same instance. This might require updating RTD configuration to Python 3 and newer Sphinx version.

**Comment by aleksihakli** [comment]
@jezdez @camilonova the work in this PR is now ready for review and possible upstreaming work-wise, I'm done with iterating over it.

There are a lot of changes in this one PR and patch, but they all relate to CI, testing, and versioning. Splitting them up doesn't really simplify things, as it's easiest to see and review their correlation at once. 

All the changes aside from the `setuptools_scm` and versioning tune-up are required or nice-to-have for a functioning Python 3 test and build setup for the project.

The docs were also broken on Python 3 due to the use of `xrange`, and I decided it would be nice to fix them in the same instance. Due to the fixes it might necessary to update RTD configuration to Python 3 and newer Sphinx version. I tested building the docs with Python 3.7 and Sphinx 2.3 and fixed a few warnings in the process.

File: setup.py [hunk 13]
```diff
@@ -1,8 +1,7 @@
-# -*- encoding: utf8 -*-
 from setuptools import setup, find_packages
 from setuptools.command.test import test
 
-from sorl import __version__, __author__, __maintainer__, __email__, __license__
+from sorl import __author__, __maintainer__, __email__, __license__
```
Comment (jezdez) [inline setup.py]: Let's remove this import completely and move all metadata inside this file (as preparation for stop using setup.py at some point completely).

File: setup.py (hunk 7 above)
Comment (jezdez) [reply setup.py]: Agreed. Let's add a [python_requires](https://packaging.python.org/guides/distributing-packages-using-setuptools/#python-requires) parameter so pip will install the right version.

File: sorl/__init__.py [hunk 14]
```diff
@@ -5,7 +5,6 @@
 
 __author__ = "Mikko Hellsing"
 __license__ = "BSD"
-__version__ = '12.6.0'
```
Comment (jezdez) [inline sorl/__init__.py]: Let's also use the `pkg_resources.get_distribution` method like in `docs/conf.py` in the `sorl/__init__.py` for the `__version__` variable?

File: sorl/__init__.py [hunk 15]
```diff
@@ -5,7 +5,6 @@
...
 __maintainer__ = "Jazzband"
 __email__ = "mariocesar@humanzilla.com"
```
Comment (jezdez) [inline sorl/__init__.py]: Let's use `roadies@jazzband.co` here.

File: .travis.yml (hunk 3 above)
Comment (jezdez) [reply .travis.yml]: Cool!

File: .travis.yml (hunk 12 above)
Comment (jezdez) [reply .travis.yml]: Nice, thank you!

File: setup.py (hunk 13 above)
Comment (aleksihakli) [reply setup.py]: @jezdez do you mean let's move all the metadata

1) from __init__.py to setup.py or
2) from setup.py to __init__.py?

I would like to migrate from using setup.py to pyproject.toml as per PEP518 in Jazzband projects at some point in time, but that will be larger migration and not in the scope of this PR :)

File: sorl/__init__.py (hunk 14 above)
Comment (aleksihakli) [reply sorl/__init__.py]: Do we want both `setup.py` and `sorl/__init__.py`? `__version__` is quasi standard anyway so both are, well, somewhat correct:

https://packaging.python.org/guides/single-sourcing-package-version/

File: setup.py (hunk 13 above)
Comment (jezdez) [reply setup.py]: The first, for the same reason as you mention. setuptools_scm [nowadays supports pyproject.toml files](https://github.com/pypa/setuptools_scm/#pyprojecttoml-usage), but it's a bit too early to adopt it just yet. Let's stay with setup.py for a bit more and let the early bugs with it be figured out.

Comment (aleksihakli) [reply setup.py]: Sure, the first was what I was thinking too, just wanted to make sure.

**Comment by aleksihakli** [comment]
@jezdez requested changes are implemented now.

File: sorl/__init__.py (hunk 14 above)
Comment (jezdez) [reply sorl/__init__.py]: Yeah, we want both, please re-add the __version__ parameter as described here: https://github.com/pypa/setuptools_scm/#retrieving-package-version-at-runtime

**Comment by jezdez** [comment]
@aleksihakli Sorry, this needs another pass for the `sorl.__version__` parameter.

**Comment by aleksihakli** [comment]
@jezdez `sorl.__version__` is now correctly configured.

File: setup.py (hunk 7 above)
Comment (benjaoming) [reply setup.py]: This does not deprecate support - AFAIK, this removes support, as pip will not find the release anymore on PyPi.

Comment (jezdez) [reply setup.py]: Yeah, I think removing was the intention given that 3.4 has already reached EOL and 3.5 is impracticable to support.

**Comment by benjaoming** [comment]
I forgot to say thanks for the work to @aleksihakli - agree on @jezdez's reasoning to remove 3.4 and 3.5 support, just think it'd be best to separate, so that the previously tagged 0.12.6 patch release (adding Django 3.0 support) could be released without the backwards incompatible and non-warned parts.

Maybe someone can upload 0.12.6 manually?

I think the PR description could be upgraded so a patch release isn't let out that removes Python 3.4 and 3.5 support.

Seems ideal for a 0.13 release, and perhaps most people would have their problems fixed if 0.13 could be immediately released from this PR?

CC: @camilonova

**Comment by aleksihakli** [comment]
@benjaoming it's probably easier to just release 0.13 as that's most sane release scheme according to Semantic Versioning https://semver.org/?

I think that 0.12.6 can just be skipped if it's not vital to release it? If it's not uploaded to PyPI then there's no harm done :)

**Comment by benjaoming** [comment]
@aleksandar-todorovic there are probably thousands of projects dependent on sorl-thumbnail, a few of which might be reusable apps trying to introduce Django 3.0 support in a complicated/diverse set of Django and Python support and release cycles.

Django 2.2 is still active, and so is Python 3.5.

If sorl-thumbnail skips 0.12.6 and removes Python 3.4 and 3.5 support, then we either have to go and create nitty-gritty conditional dependencies in `setup.py` or we have to drop our own Python 3.5 support out of the blue. This just adds more work for others with no gains.

Since there were no changes in the code, and the PR description says

> Deprecate explicit support for Python 3.4 and 3.5 in order to simplify the test matrix.

...I think the correct thing to do would be to reinstate Python 3.4 and 3.5 support in `setup.py` and release these changes as 0.12.7

This could well be the final release in the 0.12 series before moving on to a 0.13 series w/o Py 3.4 and 3.5?

**Comment by jezdez** [comment]
> @aleksandar-todorovic there are probably thousands of projects dependent on sorl-thumbnail, a few of which might be reusable apps trying to introduce Django 3.0 support in a complicated/diverse set of Django and Python support and release cycles.

For the record, the number of apps that refer to [sorl-thumbnail as a dependency is 15](https://libraries.io/pypi/sorl-thumbnail/dependents) according to libraries.io.

> 
> Django 2.2 is still active, and so is Python 3.5.
> 
> If sorl-thumbnail skips 0.12.6 and removes Python 3.4 and 3.5 support, then we either have to go and create nitty-gritty conditional dependencies in `setup.py` or we have to drop our own Python 3.5 support out of the blue. This just adds more work for others with no gains.
> 
> Since there were no changes in the code, and the PR description says
> 
> > Deprecate explicit support for Python 3.4 and 3.5 in order to simplify the test matrix.
> 
> ...I think the correct thing to do would be to reinstate Python 3.4 and 3.5 support in `setup.py` and release these changes as 0.12.7
> 
> This could well be the final release in the 0.12 series before moving on to a 0.13 series w/o Py 3.4 and 3.5?

**Comment by aleksihakli** [comment]
@benjaoming if you want Python 3.4 and 3.5 supported please feel free to add them to the package identifiers and test matrix!

I don't have much opinions on it other than Python 3.4 is officially EOL.

Python 3.5 is also decreasing in usage and I have to say that the people who run old Python versions are rarely running the latest library or framework versions.

However, adding back the identifiers is easy enough, please do so if you feel that is the best solution. I am not the maintainer, I just try to actively improve Jazzband QA automation and dream of deprecating Python versions prior to 3.6 due to feature parity :)

**Comment by benjaoming** [comment]
Sounds good @aleksandar-todorovic, thanks for fixing all this!

Sorry, I have been writing the wrong version numbers consistently in this thread! 12.7 would be a fine release according to semver for the dropped Python support. I think 12.6.1 would a be a correct bump for a release with these fixes, apart from the dropped Python support.

I'll add a PR nominating master to move to a 12.6.1 release as a step before 12.7.

**Comment by aleksihakli** [comment]
@benjaoming roger that, although I'm not Aleksandar.

**Comment by benjaoming** [comment]
haha, no, it was the first "aleks" auto-suggestion for `@` -- Github still fails to put OP first :/

**Comment by camilonova** [comment]
@aleksihakli @jezdez love you folks ❤️

**Pull request merged**